### PR TITLE
feat: add branch name issue extraction feature

### PR DESCRIPTION
**Pull request by damnthonyy**
## Summary
Add extraction of GitHub issue numbers from the PR source branch name (e.g. `feature/XX-test-issue` → issue #XX) and merge them with tickets from the PR description for compliance and review.

## Scope
- Branch-name issue extraction and `branch_issue_regex` apply only when the **git provider is GitHub**.
- Not used for Azure DevOps (linked work items) or for branch-based issue extraction on other git providers.
- Independent of how PR-Agent is triggered (GitHub Actions, Slack, API, (...) ) as long as the repo is on GitHub.

## What's done
- [x] Config: `extract_issue_from_branch`, `branch_issue_regex` in configuration.toml
- [x] `BRANCH_ISSUE_PATTERN` (option A: number at start of segment after `/` or start of branch)
- [x] `extract_tickets_link_from_branch_name()` with config and optional custom regex
- [x] Integration in `extract_tickets()` for GithubProvider (merge description + branch tickets, limit 3)
- [x] Unit tests: `tests/unittest/test_extract_issue_from_branch.py`
- [x] Manual test on dedicated repo: [pr-agent-test PR #4](https://github.com/damnthonyy/pr-agent-test/pull/4) (branch `feature/3-add-todo-items` → [pr-agent-test issue 3](https://github.com/damnthonyy/pr-agent-test/issues/3) linked from branch name without "Fixes 3" in description)

## What's left (if any)
- [ ] **Option B to explore**: support for issue number surrounded by hyphens (e.g. `something-XX-test`, `feature-42-fix`) via an additional pattern or config.
- [x] Update README or docs (short note on branch-based issue linking when enabled)
- [x] Final review / cleanup before marking ready for review

## How to test
1. Set `config.extract_issue_from_branch = true` (default).
2. Open a PR from a branch like `feature/XX-description` (issue #XX must exist in the repo if you want to test compliance).
3. Run `/review` (or CLI `review`) and confirm the ticket compliance section references issue #XX even without "Fixes #XX" in the description.
4. Run: `PYTHONPATH=. pytest tests/unittest/test_extract_issue_from_branch.py -v`
5. **Example**: [pr-agent-test PR #4](https://github.com/damnthonyy/pr-agent-test/pull/4) — branch `feature/3-add-todo-items`; review shows [pr-agent-test issue](https://github.com/damnthonyy/pr-agent-test/issues/3) in ticket compliance with no description.

> doc about the pro tip [here](https://github.blog/changelog/2022-03-01-create-a-branch-for-an-issue/)

___

### **PR Type**
Enhancement

___

### **Description**
- Add branch name issue extraction for GitHub PRs
  - Extract issue numbers from branch names (e.g., `feature/1-test-issue` → issue #1)
  - Support custom regex patterns via `branch_issue_regex` config
  - Merge branch-derived tickets with description tickets (limit 3 total)

- Add comprehensive unit tests for branch extraction logic

- Document GitHub-only limitation and usage in docs

- Add configuration options `extract_issue_from_branch` and `branch_issue_regex`

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["PR Branch Name"] -->|extract_tickets_link_from_branch_name| B["Issue Numbers"]
  C["PR Description"] -->|extract_ticket_links_from_pr_description| D["Issue Numbers"]
  B -->|merge & deduplicate| E["Combined Tickets"]
  D -->|merge & deduplicate| E
  E -->|limit to 3| F["Final Ticket List"]
  G["Config: extract_issue_from_branch<br/>branch_issue_regex"] -.->|controls| B
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>ticket_pr_compliance_check.py</strong><dd><code>Add branch name issue extraction with custom regex</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/tools/ticket_pr_compliance_check.py

<ul><li>Add <code>BRANCH_ISSUE_PATTERN</code> regex to match issue numbers at branch start <br>or after <code>/</code><br> <li> Implement <code>extract_tickets_link_from_branch_name()</code> function with config <br>and custom regex support<br> <li> Integrate branch ticket extraction into <code>extract_tickets()</code> for GitHub <br>provider<br> <li> Merge description and branch tickets with deduplication and 3-ticket <br>limit<br> <li> Add error handling for invalid custom regex patterns</ul>

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/pr-agent/pull/2231/files#diff-2a61c83755582fac0759462e8c0b9aaec02d4479e0bed047a0287caf2f3032c5">+48/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_extract_issue_from_branch.py</strong><dd><code>Comprehensive unit tests for branch extraction</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unittest/test_extract_issue_from_branch.py

<ul><li>Add 10 comprehensive unit tests for branch name extraction<br> <li> Test various branch name formats (e.g., <code>feature/1-test</code>, <code>123-fix</code>)<br> <li> Test edge cases (empty/None branch, no digits, trailing slashes)<br> <li> Test config disable behavior and invalid regex handling<br> <li> Test deduplication of multiple matches</ul>

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/pr-agent/pull/2231/files#diff-e2dbbe9b6d7a96435475910d1225197d32a98dd51c45c06490a153e4c575dd93">+97/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>fetching_ticket_context.md</strong><dd><code>Document GitHub-only branch name linking</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/docs/core-abilities/fetching_ticket_context.md

<ul><li>Add note clarifying branch-name issue linking is GitHub-only<br> <li> Document branch name format examples (e.g., <code>123-fix-bug</code>)<br> <li> Explain that branch detection applies only to GitHub provider<br> <li> Indicate support for other platforms is planned</ul>

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/pr-agent/pull/2231/files#diff-ede6b4a0dac5ca9f864dcab50e7f457535db345890a84c33fc3f727e4d4e0f0c">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>configuration.toml</strong><dd><code>Add branch extraction configuration options</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/settings/configuration.toml

<ul><li>Add <code>extract_issue_from_branch</code> config (default: true) to enable/disable <br>feature<br> <li> Add <code>branch_issue_regex</code> config for optional custom regex patterns<br> <li> Add documentation comments explaining both settings and GitHub-only <br>limitation<br> <li> Add <code>inline_file_summary</code> and <code>enable_large_pr_handling</code> config options</ul>

</details>

  </td>
  <td><a href="https://github.com/qodo-ai/pr-agent/pull/2231/files#diff-66cfda5143e484ee53ecf7aa0df7dca8ad0b181256f4b0675905db35bcbbae78">+11/-1</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: pr_agent/settings/configuration.toml [hunk 1]
```diff
@@ -110,9 +118,9 @@ collapsible_file_list_threshold=6
 inline_file_summary=false # false, true, 'table'
 # markers
 use_description_markers=false
+enable_large_pr_handling = false
```
Comment (naorpeled) [inline pr_agent/settings/configuration.toml]: Please revert this change as it's unrelated to the PR 🙏

File: .coverage
Comment (naorpeled) [inline .coverage:1]: Please remove this file 🙏

**Comment by naorpeled** [comment]
Thanks @damnthonyy 🔥
